roachpb: add gc hint replica state field and let delete range set it

aliher1911 opened the pull request:
This commit adds a test that peforms range split and merge during
cluster upgrade. This is covering test cases where raft state gets
additional information in new versions and we must ensure that
replicas don't diverge in the process of upgrade.

Release justification: Commit adds extra tests only.
Release note: None

----

When cmd_delete_range deletes the whole range it can give a GC hint so
that mvcc gc could optimize large number ranges that would be collected
using gc clear range requests.
Processing those requests together in a quick succession would reduce
compaction load on pebble.

Release justification: This change is safe as it adds handling for the new
type of request that doesn't interact with any existing functionality.
Release note: None

----

This commit adds a version gate GCHintInReplicaState for GC Hint. This
is needed to avoid divergence of replicas when older followers don't yet 
update state with a new field value.

Release justification: This commit adds a version gate for a new feature
Release note: None

----

This commit adds a gc hint field to replica state that is backed by a
replicated range local key. This field could be set explicitly by client
to indicate that all data in the range is ripe for deletion at hint
timestamp. MVCC GC could use this information to optimize clear range
requests.

Release justification: this commit is safe because it adds a new request
type and a new status field that doesn't interfere with existing
functionality.
Release note: None

aliher1911 committed:
roachpb: add gc hint replica state field

This commit adds a gc hint field to replica state that is backed by a
replicated range local key. This field could be set explicitly by client
to indicate that all data in the range is ripe for deletion at hint
timestamp. MVCC GC could use this information to optimize clear range
requests.

Release justification: this commit is safe because it adds a new request
type and a new status field that doesn't interfere with existing
functionality.
Release note: None

aliher1911 committed:
batcheval: add optional gc hint to delete_range requests

When cmd_delete_range deletes the whole range it can give a GC hint so
that mvcc gc could optimize large number ranges that would be collected
using gc clear range requests.
Processing those requests together in a quick succession would reduce
compaction load on pebble.

Release justification: This change is safe as it adds handling for the new
type of request that doesn't interact with any existing functionality.
Release note: None

erikgrinaker commented on pkg/kv/kvserver/kvserverpb/state.proto:
```diff
@@ -138,6 +138,12 @@ message ReplicaState {
+  // timestamp is set that means all data in the range is expected to be
+  // garbage at that time and MVCC GC should optimize its removal togther
+  // with other related ranges to reduce load on pebble.
+  roachpb.GCRangeHint gc_range_hint = 15 [(gogoproto.customname) = "GCRangeHint"];
```
Since the key and state methods are called `GCHint`, let's call this field and type `GCHint` as well. It's more general.

erikgrinaker commented on pkg/keys/constants.go:
```diff
@@ -92,6 +92,8 @@ var (
 	// LocalRaftTruncatedStateSuffix for the corresponding unreplicated
 	// RaftTruncatedState.
 	_ = []byte("rftt")
+	// LocalRangeGCHintSuffix is the suffix for the GC hint timestamp.
```
nit: the GC hint isn't necessarily just a timestamp, since the struct can be extended.

erikgrinaker commented on pkg/kv/kvserver/batcheval/cmd_delete_range.go:
```diff
@@ -150,9 +159,25 @@ func DeleteRange(
+		// If GCHint was provided, then we need to check if this request meets range
+		// gc criteria of removing all data.
+		if args.GcRangeHint != nil {
+			if args.Key.Equal(desc.StartKey.AsRawKey()) && args.EndKey.Equal(desc.EndKey.AsRawKey()) {
```
You'll want to do this earlier, in the  `MVCCDeleteRangeUsingTombstone` branch. This branch here with `MVCCPredicateDeleteRange` may only remove parts of the range data.

erikgrinaker commented on pkg/roachpb/metadata.go:
```diff
@@ -773,3 +774,28 @@ func (l Locality) AddTier(tier Tier) Locality {
+func (h *GCRangeHint) Merge(rhs *GCRangeHint) bool {
+	leftEmpty := h.EstimatedRangeRemovalTime.IsEmpty()
+	rightEmpty := rhs.EstimatedRangeRemovalTime.IsEmpty()
+	if leftEmpty || rightEmpty {
+		// If one of sides doesn't have hint we probably can't do anything sensible.
```
Would it be better to just use the non-empty one?

aliher1911 replied:
My initial thinking was that if we merge with a range that had no hint it means it will likely have data in it so it would be a waste. But maybe you are right and we can still run the range through expedited checks and probe if they are still eligible if we merged "hinted" ranges with empty ones on the range boundaries.

erikgrinaker commented on pkg/roachpb/api.proto:
```diff
@@ -384,6 +384,10 @@ message DeleteRangeRequest {
+  // If set and DeleteRangeRequest is deleting the whole range with a range
+  // tombstone, gc_range_hint will update replica GC hint to notify mvcc gc
+  // queue that this range would benefit from optimized GC process.
+  GCRangeHint gc_range_hint = 8;
```
I wouldn't send the entire GC hint explicitly here, since we may not want callers to get access to everything in that struct. I'd add a boolean field, something like `AggressiveGC` or `IsSchemaDrop` or something, and then set the GC hint to the current timestamp based on that.

erikgrinaker commented on pkg/roachpb/metadata.proto:
```diff
@@ -451,3 +451,12 @@ message Version {
+
+  // estimated_range_removal_time indicates projected time when all data in the range
+  // would become garbage. If empty, no such estimate is known.
+  optional util.hlc.Timestamp estimated_range_removal_time = 2 [(gogoproto.nullable) = false];
```
Does this mean that we'd calculate the GC TTL and set the timestamp when the range would fall below the GC TTL? What if the user changes the GC TTL after the drop, to GC it faster?

I think we'll want to set this to the `DeleteRange` evaluation timestamp, and then the GC queue can compare that to the GC TTL.

erikgrinaker reviewed (commented):
Initial draft pass. There are also a few nits here and there, but will wait for the final PR.

tbg reviewed (dismissed):
I think we need a migration here, or we risk divergence in mixed-version clusters where some nodes write the hint during merge/split and others don't.

That is, we need to encode, into the split/merge trigger, the information on whether GCHint can be written. We need a cluster version to gate this during evaluation, and only if that version is active do we inform the commit trigger to act on GCHints. In particular we need to be careful about `WriteInitialReplicaState` which is called from splits.

Basic approach looks good!

> Reviewed 1 of 14 files at r1, 17 of 17 files at r3, 6 of 6 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/86565#-:-NA9YIgF0oBocFmEWjJr:b2r2sj5)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aliher1911 and @erikgrinaker)

___
*[`pkg/keys/keys.go` line 283 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/86565#-NA9WZYL2dhEemIlvDQs:-NA9WZYL2dhEemIlvDQt:b-myae33) ([raw file](https://github.com/cockroachdb/cockroach/blob/feae0e570277d369adf52a62d524a4777e838a16/pkg/keys/keys.go#L283)):*
> ```Go
> 
> // RangeGCHintKey returns a system-local key for last used GC threshold on the
> // user keyspace. Reads and writes <= this timestamp will not be served.
> ```

Copy pasta error
___
*[`pkg/kv/kvserver/batcheval/cmd_delete_range.go` line 62 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/86565#-NA9Yt170igAOGOkLVIR:-NA9Yt170igAOGOkLVIS:bj82ouq) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ac298e3bfe9f7b9caf093a9f96f762a0eee39a7/pkg/kv/kvserver/batcheval/cmd_delete_range.go#L62)):*
> ```Go
> 		})
> 
> 		if args.UpdateDeleteRangeGcHint {
> ```

GCHint (use `gogoproto.customname`)
___
*[`pkg/kv/kvserver/batcheval/cmd_delete_range.go` line 93 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/86565#-NA9YyQS-Y3nOwOdStwT:-NA9YyQS-Y3nOwOdStwU:b32vtt) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ac298e3bfe9f7b9caf093a9f96f762a0eee39a7/pkg/kv/kvserver/batcheval/cmd_delete_range.go#L93)):*
> ```Go
> 		// usage of hint.
> 		return result.Result{}, errors.AssertionFailedf(
> 			"GcRangeHint must only be used together with UseRangeTombstone")
> ```

GC
___
*[`pkg/kv/kvserver/batcheval/cmd_delete_range.go` line 112 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/86565#-NA9Z0-95bZLa_Hhooe8:-NA9Z0-95bZLa_Hhooe9:b-2v27ug) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ac298e3bfe9f7b9caf093a9f96f762a0eee39a7/pkg/kv/kvserver/batcheval/cmd_delete_range.go#L112)):*
> ```Go
> 
> 		maybeUpdateGCHint := func(res *result.Result) error {
> 			if args.UpdateDeleteRangeGcHint {
> ```

You can return early to make this less indented.

```go
if !args.UpdateDeleteRangeGCHint {
  return nil
}
if !args.Key.Equal(...) {
  return nil
}
...
```
___
*[`pkg/kv/kvserver/batcheval/cmd_delete_range.go` line 132 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/86565#-NA9Z9ZA4-Rt62UvLM59:-NA9Z9ZA4-Rt62UvLM5A:b-qn0owu) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ac298e3bfe9f7b9caf093a9f96f762a0eee39a7/pkg/kv/kvserver/batcheval/cmd_delete_range.go#L132)):*
> ```Go
> 					}
> 				} else {
> 					log.Warningf(ctx, "attempt to set GC hint for the incompletely covered range")
> ```

This can happen legitimately, so I don't think it should log as warning. I think it's generally not a good idea to log from `batcheval`. I would just remove this. Instead, we could just not write the hint unless the entire range is affected.
___
*[`pkg/kv/kvserver/batcheval/cmd_end_transaction.go` line 1234 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/86565#-NA9Xiji3u--56YxSHsD:-NA9Xiji3u--56YxSHsE:b1u0j88) ([raw file](https://github.com/cockroachdb/cockroach/blob/feae0e570277d369adf52a62d524a4777e838a16/pkg/kv/kvserver/batcheval/cmd_end_transaction.go#L1234)):*
> ```Go
> 	{
> 		// If we have GC hints populated that means we are trying to perform
> 		// optimized garbage removal in future. For load based splits we want to
> ```

I don't understand what load based splits have to do with this.
___
*[`pkg/kv/kvserver/batcheval/cmd_end_transaction.go` line 1237 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/86565#-NA9Xr8556Ij73id1AMv:-NA9Xr8556Ij73id1AMw:b-az2l5f) ([raw file](https://github.com/cockroachdb/cockroach/blob/feae0e570277d369adf52a62d524a4777e838a16/pkg/kv/kvserver/batcheval/cmd_end_transaction.go#L1237)):*
> ```Go
> 		// preserve this behaviour as both ranges likely belong to the same table.
> 		// We will try to merge both hints if possible and set new hint on LHS.
> 		// If resulting hint makes no sense, merge will reset it so that it is reset
> ```

what does "makes no sense" mean? I think we'd just track the larger of the two hints. If it isn't as easy as that explain in more words.
___
*[`pkg/kv/kvserver/batcheval/result/result.go` line 254 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/86565#-NA9WpdZBfm1iFkwLRSs:-NA9WpdZBfm1iFkwLRSt:b-4zm22g) ([raw file](https://github.com/cockroachdb/cockroach/blob/feae0e570277d369adf52a62d524a4777e838a16/pkg/kv/kvserver/batcheval/result/result.go#L254)):*
> ```Go
> 
> 		if q.Replicated.State.GCHint != nil {
> 			p.Replicated.State.GCHint = q.Replicated.State.GCHint
> ```

Detect conflicting GCHints here to catch potential programming errors.
___
*[`pkg/kv/kvserver/stateloader/initial.go` line 97 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/86565#-NA9XCFt-u3zi_I1I5Hs:-NA9XCFt-u3zi_I1I5Ht:b-qmwx02) ([raw file](https://github.com/cockroachdb/cockroach/blob/feae0e570277d369adf52a62d524a4777e838a16/pkg/kv/kvserver/stateloader/initial.go#L97)):*
> ```Go
> 		return enginepb.MVCCStats{}, errors.Wrap(err, "error reading GCHint")
> 	} else if !existingGCHint.LatestRangeDeleteTimestamp.IsEmpty() {
> 		log.Fatalf(ctx, "expected trivial GCHint, but found %+v", existingGCHint)
> ```

return an `errors.AssertionFailedf`
___
*[`pkg/kv/kvserver/stateloader/stateloader.go` line 283 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/86565#-NA9XIoUCQ4VQHGD7kZy:-NA9XIoUCQ4VQHGD7kZz:b3kqx) ([raw file](https://github.com/cockroachdb/cockroach/blob/feae0e570277d369adf52a62d524a4777e838a16/pkg/kv/kvserver/stateloader/stateloader.go#L283)):*
> ```Go
> }
> 
> // LoadGCHint loads GC hint
> ```

.
___
*[`pkg/kv/kvserver/stateloader/stateloader.go` line 293 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/86565#-NA9XJXND_lDuo8AbR3l:-NA9XJXND_lDuo8AbR3m:b3kqx) ([raw file](https://github.com/cockroachdb/cockroach/blob/feae0e570277d369adf52a62d524a4777e838a16/pkg/kv/kvserver/stateloader/stateloader.go#L293)):*
> ```Go
> }
> 
> // SetGCHint writes the GC hint
> ```

.
___
*[`pkg/roachpb/metadata.go` line 786 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/86565#-NA9XSVODW79s1RszXHC:-NA9XSVODW79s1RszXHD:bi8m9pf) ([raw file](https://github.com/cockroachdb/cockroach/blob/feae0e570277d369adf52a62d524a4777e838a16/pkg/roachpb/metadata.go#L786)):*
> ```Go
> // Returns true if receiver state was changed.
> func (h *GCRangeHint) Merge(rhs *GCRangeHint) bool {
> 	if h.LatestRangeDeleteTimestamp.Less(rhs.LatestRangeDeleteTimestamp) {
> ```

isn't there a method `.Forward` for this?
___
*[`pkg/roachpb/metadata.proto` line 461 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/86565#-NA4t3wnhlr0r8ozXk8v-r2-461:-NA9X_Tz3t0JhVSyHIay:b-nh7r4e) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e60dad3d96cdb60e1a7224a9e3ac5b1b575d9d7/pkg/roachpb/metadata.proto#L461)):*
<details><summary><i>Previously, erikgrinaker (Erik Grinaker) wrote…</i></summary><blockquote>

Does this mean that we'd calculate the GC TTL and set the timestamp when the range would fall below the GC TTL? What if the user changes the GC TTL after the drop, to GC it faster?

I think we'll want to set this to the `DeleteRange` evaluation timestamp, and then the GC queue can compare that to the GC TTL.
</blockquote></details>

+1 let's set this to the write timestamp during eval


<!-- Sent from Reviewable.io -->

aliher1911 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/86565)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @erikgrinaker and @tbg)

___
*[`pkg/kv/kvserver/batcheval/result/result.go` line 254 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/86565#-NA9WpdZBfm1iFkwLRSs:-NA9ggtD41PAQ6s6ou2_:bpe5c1r) ([raw file](https://github.com/cockroachdb/cockroach/blob/feae0e570277d369adf52a62d524a4777e838a16/pkg/kv/kvserver/batcheval/result/result.go#L254)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Detect conflicting GCHints here to catch potential programming errors.
</blockquote></details>

Should we always reject if it is set on both sides of merge? What would be the use case, sending multiple identical delete requests in a batch?
___
*[`pkg/kv/kvserver/stateloader/initial.go` line 97 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/86565#-NA9XCFt-u3zi_I1I5Hs:-NA9fWy9BnzQRWUUK7UK:bx8bi1m) ([raw file](https://github.com/cockroachdb/cockroach/blob/feae0e570277d369adf52a62d524a4777e838a16/pkg/kv/kvserver/stateloader/initial.go#L97)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

return an `errors.AssertionFailedf`
</blockquote></details>

Why do other checks here resort to log.Fatalf?
___
*[`pkg/roachpb/api.proto` line 390 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/86565#-NA4t3wnhlr0r8ozXk8u-r2-390:-NA9eTmx9bv4khdbV3ne:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e60dad3d96cdb60e1a7224a9e3ac5b1b575d9d7/pkg/roachpb/api.proto#L390)):*
<details><summary><i>Previously, erikgrinaker (Erik Grinaker) wrote…</i></summary><blockquote>

I wouldn't send the entire GC hint explicitly here, since we may not want callers to get access to everything in that struct. I'd add a boolean field, something like `AggressiveGC` or `IsSchemaDrop` or something, and then set the GC hint to the current timestamp based on that.
</blockquote></details>

Done.
___
*[`pkg/roachpb/metadata.go` line 786 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/86565#-NA9XSVODW79s1RszXHC:-NA9bjc56CHOAm5MTPEK:bmsy8yh) ([raw file](https://github.com/cockroachdb/cockroach/blob/feae0e570277d369adf52a62d524a4777e838a16/pkg/roachpb/metadata.go#L786)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

isn't there a method `.Forward` for this?
</blockquote></details>

Forward is weird because it changes synthetic flag in timestamp but returns false indicating that it didn't forward it. I would be scared to have equal objects with different content in the state.
___
*[`pkg/roachpb/metadata.proto` line 461 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/86565#-NA4t3wnhlr0r8ozXk8v-r2-461:-NA9eYrR2zTViQmmsJty:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e60dad3d96cdb60e1a7224a9e3ac5b1b575d9d7/pkg/roachpb/metadata.proto#L461)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

+1 let's set this to the write timestamp during eval
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

aliher1911 committed:
kvserver: add version gate to gc hint in replica state

This commit adds a version gate for GC Hint. This is needed to avoid
divergence of replicas when older followers don't yet update state with
a new field value.

Release justification: This commit adds a version gate for a new feature
Release note: None

erikgrinaker reviewed (commented):
> I think we need a migration here

A below-Raft migration, or just a version gate? Was really hoping we'd avoid a migration this cycle, given the issues with the below-Raft migration infrastructure.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/86565#-:-NAAj5Gi2d_Ho8LuMW-i:b-v9ygy)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aliher1911, @erikgrinaker, and @tbg)

___
*[`pkg/kv/kvserver/batcheval/result/result.go` line 254 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/86565#-NA9WpdZBfm1iFkwLRSs:-NAAkJU-23B8Y-qFGWnq:bul14f1) ([raw file](https://github.com/cockroachdb/cockroach/blob/feae0e570277d369adf52a62d524a4777e838a16/pkg/kv/kvserver/batcheval/result/result.go#L254)):*
<details><summary><i>Previously, aliher1911 (Oleg) wrote…</i></summary><blockquote>

Should we always reject if it is set on both sides of merge? What would be the use case, sending multiple identical delete requests in a batch?
</blockquote></details>

I think we need to do something similar to `GCThreshold`, and forward it to the highest timestamp. Consider e.g. serverless, where we can have multiple dropped tables in the same range, and we may merge ranges with different dropped tables in the LHS and RHS. Forwarding to the highest timestamp seems like the most reasonable action here, even though it may end up being a futile GC attempt.


<!-- Sent from Reviewable.io -->

erikgrinaker commented on pkg/kv/kvserver/batcheval/cmd_delete_range_gchint_test.go:
```diff
@@ -0,0 +1,81 @@
+	"github.com/stretchr/testify/require"
+)
+
+func TestDeleteRangeTombstoneSetsGCHint(t *testing.T) {
```
nit: let's put this in `cmd_delete_range_test.go`, it doesn't seem significant enough to warrant its own file.

erikgrinaker reviewed (commented):
Still a bit iffy on the migration details here -- will defer to Tobi.

> Reviewed 2 of 17 files at r3, 6 of 15 files at r5, 1 of 10 files at r7, 6 of 7 files at r10, 1 of 7 files at r12, 9 of 9 files at r13, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/86565#-:-NADpGz62Cu1qTQz2VIo:bb0x310)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aliher1911 and @tbg)

___
*[`pkg/kv/kvserver/replica_command.go` line 232 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/86565#-NADklaXACTu9HfVkl8R:-NADklaXACTu9HfVkl8S:b96ivv8) ([raw file](https://github.com/cockroachdb/cockroach/blob/30d44742bcd2d0bcef0115cf5f95d8d2f248ed80/pkg/kv/kvserver/replica_command.go#L232)):*
> ```Go
> 	}
> 
> 	supoortsGCHints := store.ClusterSettings().Version.IsActive(ctx, clusterversion.GCHintInReplicaState)
> ```

`supportsGCHints`. Ditto below.
___
*[`pkg/kv/kvserver/store_snapshot.go` line 1302 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/86565#-NADlB4V7E-JdR3LUngF:-NADlB4V7E-JdR3LUngG:bo5j1o3) ([raw file](https://github.com/cockroachdb/cockroach/blob/30d44742bcd2d0bcef0115cf5f95d8d2f248ed80/pkg/kv/kvserver/store_snapshot.go#L1302)):*
> ```Go
> 		st.Version.ActiveVersionOrEmpty(ctx).Version,
> 		true, /* 22.1:AddRaftAppliedIndexTermMigration */
> 		true, /* 22.2: GCHintInReplicaState */
> ```

Don't we have to gate this on the version gate?
___
*[`pkg/kv/kvserver/batcheval/cmd_delete_range.go` line 118 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/86565#-NADp5je8vwm1Uk_8uSm:-NADp5je8vwm1Uk_8uSn:bdlbb4d) ([raw file](https://github.com/cockroachdb/cockroach/blob/30d44742bcd2d0bcef0115cf5f95d8d2f248ed80/pkg/kv/kvserver/batcheval/cmd_delete_range.go#L118)):*
> ```Go
> 			// If GCHint was provided, then we need to check if this request meets
> 			// range gc criteria of removing all data. This is not an error as range
> 			// might have split/merged since request was sent and we don't want to
> ```

nit: only merges would affect this -- a split would cause the request to get split as well.
___
*[`pkg/kv/kvserver/batcheval/cmd_delete_range.go` line 128 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/86565#-NADordT7IuYrS6QNs0B:-NADordT7IuYrS6QNs0C:b87nl4f) ([raw file](https://github.com/cockroachdb/cockroach/blob/30d44742bcd2d0bcef0115cf5f95d8d2f248ed80/pkg/kv/kvserver/batcheval/cmd_delete_range.go#L128)):*
> ```Go
> 				return err
> 			}
> 			if !hint.UpdateLatestDeleteRangeTimestamp(h.Timestamp) {
> ```

nit: Maybe `Forward`. It wasn't clear to me why this could fail without reading the method.
___
*[`pkg/kv/kvserver/batcheval/cmd_delete_range_gchint_test.go` line 28 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/86565#-NADmQdz-JALmajol9PL:-NADmQe-BZ5yToFUeEMu:b-roap8i) ([raw file](https://github.com/cockroachdb/cockroach/blob/30d44742bcd2d0bcef0115cf5f95d8d2f248ed80/pkg/kv/kvserver/batcheval/cmd_delete_range_gchint_test.go#L28)):*
> ```Go
> )
> 
> func TestDeleteRangeTombstoneSetsGCHint(t *testing.T) {
> ```

nit: let's move this into `cmd_delete_range_test.go`, it doesn't seem significant enough to deserve its own file.
___
*[`pkg/kv/kvserver/batcheval/cmd_end_transaction.go` line 1240 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/86565#-NADoF-FE08As8Z1-TDp:-NADoF-FE08As8Z1-TDq:b-1dqbgv) ([raw file](https://github.com/cockroachdb/cockroach/blob/30d44742bcd2d0bcef0115cf5f95d8d2f248ed80/pkg/kv/kvserver/batcheval/cmd_end_transaction.go#L1240)):*
> ```Go
> 		// optimized garbage removal in future.
> 		// We will try to merge both hints if possible and set new hint on LHS.
> 		lshLoader := MakeStateLoader(rec)
> ```

`lhsLoader`
___
*[`pkg/kv/kvserver/batcheval/result/result.go` line 254 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/86565#-NA9WpdZBfm1iFkwLRSs:-NADog4r91uUATZ1iKWI:bdo5q82) ([raw file](https://github.com/cockroachdb/cockroach/blob/feae0e570277d369adf52a62d524a4777e838a16/pkg/kv/kvserver/batcheval/result/result.go#L254)):*
<details><summary><i>Previously, erikgrinaker (Erik Grinaker) wrote…</i></summary><blockquote>

I think we need to do something similar to `GCThreshold`, and forward it to the highest timestamp. Consider e.g. serverless, where we can have multiple dropped tables in the same range, and we may merge ranges with different dropped tables in the LHS and RHS. Forwarding to the highest timestamp seems like the most reasonable action here, even though it may end up being a futile GC attempt.
</blockquote></details>

Right, ok, this is for merging batch results. We set `isAlone` on `UseRangeTombstone`, so this should be fine.
___
*[`pkg/roachpb/api.proto` line 389 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/86565#-NADnEnNAzaWhiDHRBiN:-NADnEnNAzaWhiDHRBiO:b-silt7i) ([raw file](https://github.com/cockroachdb/cockroach/blob/30d44742bcd2d0bcef0115cf5f95d8d2f248ed80/pkg/roachpb/api.proto#L389)):*
> ```protobuf
>   bool idempotent_tombstone = 7;
>   // If enabled and DeleteRangeRequest is deleting the whole range with a range
>   // tombstone, GCRangeHint on replica will be updated to notify mvcc gc queue
> ```

`GCHint`
___
*[`pkg/roachpb/api.proto` line 391 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/86565#-NADo-5b5KDhx06F5EFp:-NADo-5b5KDhx06F5EFq:b-agf2de) ([raw file](https://github.com/cockroachdb/cockroach/blob/30d44742bcd2d0bcef0115cf5f95d8d2f248ed80/pkg/roachpb/api.proto#L391)):*
> ```protobuf
>   // tombstone, GCRangeHint on replica will be updated to notify mvcc gc queue
>   // that this range would benefit from optimized GC process.
>   bool update_delete_range_gc_hint = 8 [(gogoproto.customname) = "UpdateDeleteRangeGCHint"];
> ```

The hint name is `range_delete`, so let's use that.
___
*[`pkg/roachpb/metadata.go` line 798 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/86565#-NADnU5U3KZrBmj-zaWl:-NADnU5U3KZrBmj-zaWm:b-csu81x) ([raw file](https://github.com/cockroachdb/cockroach/blob/30d44742bcd2d0bcef0115cf5f95d8d2f248ed80/pkg/roachpb/metadata.go#L798)):*
> ```Go
> // UpdateLatestDeleteRangeTimestamp bumps LatestDeleteRangeTimestamp in GC hint
> // if it is greater than previously set.
> func (h *GCHint) UpdateLatestDeleteRangeTimestamp(ts hlc.Timestamp) bool {
> ```

`RangeDelete`, like field name.
___
*[`pkg/roachpb/metadata.proto` line 459 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/86565#-NADntbo73XsJYiJ5Xky:-NADntbo73XsJYiJ5Xkz:bicdmjx) ([raw file](https://github.com/cockroachdb/cockroach/blob/30d44742bcd2d0bcef0115cf5f95d8d2f248ed80/pkg/roachpb/metadata.proto#L459)):*
> ```protobuf
>   option (gogoproto.equal) = true;
> 
>   // LatestRangeDeleteTimestamp indicates projected time when all data in the range
> ```

This comment isn't accurate.


<!-- Sent from Reviewable.io -->

aliher1911 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/86565)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @erikgrinaker and @tbg)

___
*[`pkg/kv/kvserver/store_snapshot.go` line 1302 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/86565#-NADlB4V7E-JdR3LUngF:-NAE6Bk34Y3PbV6oUs01:b-49gxqu) ([raw file](https://github.com/cockroachdb/cockroach/blob/30d44742bcd2d0bcef0115cf5f95d8d2f248ed80/pkg/kv/kvserver/store_snapshot.go#L1302)):*
<details><summary><i>Previously, erikgrinaker (Erik Grinaker) wrote…</i></summary><blockquote>

Don't we have to gate this on the version gate?
</blockquote></details>

Possibly? It is used in ResetQuorum op only and I'm not sure when it is being invoked tbh. I guess it won't hurt to add a version gate.
The init cluster case doesn't need that I assumed as we never initialize with "old" version anyways.
___
*[`pkg/kv/kvserver/batcheval/cmd_delete_range_gchint_test.go` line 28 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/86565#-NADmQdz-JALmajol9PL:-NAE4_Nn5CK4R_SxkEPN:b-7po1il) ([raw file](https://github.com/cockroachdb/cockroach/blob/30d44742bcd2d0bcef0115cf5f95d8d2f248ed80/pkg/kv/kvserver/batcheval/cmd_delete_range_gchint_test.go#L28)):*
<details><summary><i>Previously, erikgrinaker (Erik Grinaker) wrote…</i></summary><blockquote>

nit: let's move this into `cmd_delete_range_test.go`, it doesn't seem significant enough to deserve its own file.
</blockquote></details>

This is a bit unfortunate, but we can't use TestServer/TestCluster from batcheval because of cycling dependency. This test lives in batcheval_test for this reason. While other batch_eval tests rely on package local names.
We can expose local names and move other tests into batcheval_test package, not sure it is a good idea though.


<!-- Sent from Reviewable.io -->

erikgrinaker reviewed (commented):
> Reviewed 17 of 17 files at r14, 6 of 7 files at r15, 10 of 10 files at r17, 9 of 9 files at r18, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/86565)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aliher1911 and @tbg)

___
*[`pkg/kv/kvserver/store_snapshot.go` line 1302 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/86565#-NADlB4V7E-JdR3LUngF:-NAEA1Td8RTBGRQsRHHS:bamyzas) ([raw file](https://github.com/cockroachdb/cockroach/blob/30d44742bcd2d0bcef0115cf5f95d8d2f248ed80/pkg/kv/kvserver/store_snapshot.go#L1302)):*
<details><summary><i>Previously, aliher1911 (Oleg) wrote…</i></summary><blockquote>

Possibly? It is used in ResetQuorum op only and I'm not sure when it is being invoked tbh. I guess it won't hurt to add a version gate.
The init cluster case doesn't need that I assumed as we never initialize with "old" version anyways.
</blockquote></details>

`AddRaftAppliedIndexTermMigration` used to error in this case, that seems prudent:

https://github.com/cockroachdb/cockroach/pull/75675/commits/c3bc064a2ccc5cdf97235e25888e8d4c03d0e37d#diff-b64eaa053bd62a0257f609359e87e5060644dfd084a3768fc0e4035bb8a2819c
___
*[`pkg/kv/kvserver/batcheval/cmd_delete_range_gchint_test.go` line 28 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/86565#-NADmQdz-JALmajol9PL:-NAE8Igd36Zk-a6DjWq0:b-8my34e) ([raw file](https://github.com/cockroachdb/cockroach/blob/30d44742bcd2d0bcef0115cf5f95d8d2f248ed80/pkg/kv/kvserver/batcheval/cmd_delete_range_gchint_test.go#L28)):*
<details><summary><i>Previously, aliher1911 (Oleg) wrote…</i></summary><blockquote>

This is a bit unfortunate, but we can't use TestServer/TestCluster from batcheval because of cycling dependency. This test lives in batcheval_test for this reason. While other batch_eval tests rely on package local names.
We can expose local names and move other tests into batcheval_test package, not sure it is a good idea though.
</blockquote></details>

I see, that's fine.


<!-- Sent from Reviewable.io -->

aliher1911 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/86565)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @erikgrinaker and @tbg)

___
*[`pkg/kv/kvserver/store_snapshot.go` line 1302 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/86565#-NADlB4V7E-JdR3LUngF:-NAEBX6A8gRVzvne6JhX:b-pi3y01) ([raw file](https://github.com/cockroachdb/cockroach/blob/30d44742bcd2d0bcef0115cf5f95d8d2f248ed80/pkg/kv/kvserver/store_snapshot.go#L1302)):*
<details><summary><i>Previously, erikgrinaker (Erik Grinaker) wrote…</i></summary><blockquote>

`AddRaftAppliedIndexTermMigration` used to error in this case, that seems prudent:

https://github.com/cockroachdb/cockroach/pull/75675/commits/c3bc064a2ccc5cdf97235e25888e8d4c03d0e37d#diff-b64eaa053bd62a0257f609359e87e5060644dfd084a3768fc0e4035bb8a2819c
</blockquote></details>

I'll add an assertion there.


<!-- Sent from Reviewable.io -->

tbg reviewed (commented):
> Reviewed 5 of 15 files at r5, 1 of 10 files at r7, 17 of 17 files at r14, 6 of 7 files at r15, 10 of 10 files at r19, 9 of 9 files at r20, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/86565)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aliher1911)

___
*[`-- commits` line 19 at r19](https://reviewable.io/reviews/cockroachdb/cockroach/86565#-NAEuNib-h2-EnVEfLyJ):*
nit: divergence
___
*[`pkg/clusterversion/cockroach_versions.go` line 297 at r19](https://reviewable.io/reviews/cockroachdb/cockroach/86565#-NAEuS7qAXLjsCwSaVkD:-NAEuS7qAXLjsCwSaVkE:b-hsqqmo) ([raw file](https://github.com/cockroachdb/cockroach/blob/14b5ca8aaad023558d2a02bbdc48ddabdd2850d6/pkg/clusterversion/cockroach_versions.go#L297)):*
> ```Go
> 	// AddSSTable calls will be used outside of tenant streaming.
> 	NoNonMVCCAddSSTable
> 	// GCHintInReplicaState adds GC hint to replica state. When this option is
> ```

s/option/version/
___
*[`pkg/kv/kvserver/batcheval/cmd_delete_range.go` line 128 at r20](https://reviewable.io/reviews/cockroachdb/cockroach/86565#-NAEvc0X5zN2Jp8X_ruM:-NAEvc0X5zN2Jp8X_ruN:b-h4oova) ([raw file](https://github.com/cockroachdb/cockroach/blob/2ee1f7caa4f2f189daabfb3318704c2ec4dbcb22/pkg/kv/kvserver/batcheval/cmd_delete_range.go#L128)):*
> ```Go
> 				return err
> 			}
> 			if !hint.ForwardLatestRangeDeleteTimestamp(h.Timestamp) {
> ```

is `h.Timestamp` right? Don't you want `h.WriteTimestamp()`?
___
*[`pkg/kv/kvserver/batcheval/result/result.go` line 254 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/86565#-NA9WpdZBfm1iFkwLRSs:-NAEtIfW-8gcr1VGdd3a:b-hoco7v) ([raw file](https://github.com/cockroachdb/cockroach/blob/feae0e570277d369adf52a62d524a4777e838a16/pkg/kv/kvserver/batcheval/result/result.go#L254)):*
<details><summary><i>Previously, erikgrinaker (Erik Grinaker) wrote…</i></summary><blockquote>

Right, ok, this is for merging batch results. We set `isAlone` on `UseRangeTombstone`, so this should be fine.
</blockquote></details>

Sorry, should've specified that it would be an assertion for now, though in the future yes we could pick the higher hint.
___
*[`pkg/kv/kvserver/stateloader/initial.go` line 97 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/86565#-NA9XCFt-u3zi_I1I5Hs:-NAEtlDf44bXyOHTStUn:b-pgrsbu) ([raw file](https://github.com/cockroachdb/cockroach/blob/feae0e570277d369adf52a62d524a4777e838a16/pkg/kv/kvserver/stateloader/initial.go#L97)):*
<details><summary><i>Previously, aliher1911 (Oleg) wrote…</i></summary><blockquote>

Why do other checks here resort to log.Fatalf?
</blockquote></details>

Bad precedent. We used to be too heavy on `log.Fatal` but the mantra should be to handle through errors what we can, to avoid needlessly taking down production services when it can be avoided.
___
*[`pkg/kv/kvserver/stateloader/stateloader.go` line 290 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/86565#-NAEtypt4xiTB3n8hr9i:-NAEtypt4xiTB3n8hr9j:bqggfw) ([raw file](https://github.com/cockroachdb/cockroach/blob/f595c74e2c91957ff2b4a2d2666726b6eed680db/pkg/kv/kvserver/stateloader/stateloader.go#L290)):*
> ```Go
> 	_, err := storage.MVCCGetProto(ctx, reader, rsl.RangeGCHintKey(),
> 		hlc.Timestamp{}, &h, storage.MVCCGetOptions{})
> 	return &h, err
> ```

nit: `if err := nil { return nil, err }; return &h, nil`
___
*[`pkg/roachpb/data.proto` line 134 at r19](https://reviewable.io/reviews/cockroachdb/cockroach/86565#-NAEvDP6AgRFSjApSd1R:-NAEvDP72IOVTG3CjeNV:b1mvtn7) ([raw file](https://github.com/cockroachdb/cockroach/blob/14b5ca8aaad023558d2a02bbdc48ddabdd2850d6/pkg/roachpb/data.proto#L134)):*
> ```protobuf
>   // WriteGCHint is a version gate to only enable writing GC hints in split
>   // trigger if cluster was upgraded to required version.
>   bool write_gc_hint = 4 [(gogoproto.customname) = "WriteGCHint"];
> ```

TODO(replication): remove this field in the 23.2 cycle (i.e. on master after 23.1 has been cut).
___
*[`pkg/roachpb/data.proto` line 215 at r19](https://reviewable.io/reviews/cockroachdb/cockroach/86565#-NAEvI6p-tqm8FTf9tGX:-NAEvI6p-tqm8FTf9tGY:b-7ropln) ([raw file](https://github.com/cockroachdb/cockroach/blob/14b5ca8aaad023558d2a02bbdc48ddabdd2850d6/pkg/roachpb/data.proto#L215)):*
> ```protobuf
>   // WriteGCHint is a version gate to only enable writing GC hints in merge
>   // trigger if cluster was upgraded to required version.
>   bool write_gc_hint = 8 [(gogoproto.customname) = "WriteGCHint"];
> ```

ditto
___
*[`pkg/roachpb/metadata.go` line 782 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/86565#-NAEu81H1ViZnU5CWnXW:-NAEu81H1ViZnU5CWnXX:b-dib0kk) ([raw file](https://github.com/cockroachdb/cockroach/blob/f595c74e2c91957ff2b4a2d2666726b6eed680db/pkg/roachpb/metadata.go#L782)):*
> ```Go
> }
> 
> // Merge combines GC hint information about merging replicas and updates
> ```

Say that it forwards `h` to `rhs` and returns true if that changed anything.


<!-- Sent from Reviewable.io -->

erikgrinaker reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/86565)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aliher1911 and @tbg)

___
*[`pkg/kv/kvserver/batcheval/cmd_delete_range.go` line 128 at r20](https://reviewable.io/reviews/cockroachdb/cockroach/86565#-NAEvc0X5zN2Jp8X_ruM:-NAEx-9HCPJB4vwWGFtz:b9fhwvg) ([raw file](https://github.com/cockroachdb/cockroach/blob/2ee1f7caa4f2f189daabfb3318704c2ec4dbcb22/pkg/kv/kvserver/batcheval/cmd_delete_range.go#L128)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

is `h.Timestamp` right? Don't you want `h.WriteTimestamp()`?
</blockquote></details>

Range tombstones are non-transactional, so this can only be `h.Timestamp`.


<!-- Sent from Reviewable.io -->

tbg commented on pkg/kv/kvserver/batcheval/cmd_end_transaction.go:
```diff
@@ -1217,6 +1232,30 @@ func mergeTrigger(
+		if err != nil {
+			return result.Result{}, err
+		}
+		if lhsHint.Merge(rhsHint) {
```
I *think* this is correct but also it's brittle? If the cluster setting hasn't been bumped yet, we'll get `false` here, because the right-hand side can't possibly have a GCHint, and so we're definitely not getting a `true` here and thus are not calling `SetGCHint`.

Looking at the way all the plumbing works now I am a little confused why we don't put the bool on `SetGCHint`. By definition, anyone writing the GCHint has to go through that method. Yet, at the moment that method will happily write a GCHint even when it's not allowed and we have to check each caller.

If the method instead took a `GCHintsEnabled bool`, every caller would have to be aware of it.

I just pushed a commit that attempts it to this PR. Mind taking a look and seeing if you prefer it? (I do).

aliher1911 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/86565)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aliher1911, @erikgrinaker, and @tbg)

___
*[`pkg/kv/kvserver/batcheval/cmd_end_transaction.go` line 1250 at r20](https://reviewable.io/reviews/cockroachdb/cockroach/86565#-NA5tWpgph4aEvJVTdUW-r20-1250:-NAFEmAsBgWuOjcy8KZR:bi8h2f6) ([raw file](https://github.com/cockroachdb/cockroach/blob/2ee1f7caa4f2f189daabfb3318704c2ec4dbcb22/pkg/kv/kvserver/batcheval/cmd_end_transaction.go#L1250)):*
> I just pushed a commit that attempts it to this PR. Mind taking a look and seeing if you prefer it? (I do).

Looks good, I'll merge that approach appropriately into the commits.


<!-- Sent from Reviewable.io -->

aliher1911 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/86565)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @erikgrinaker and @tbg)

___
*[`pkg/kv/kvserver/batcheval/cmd_delete_range.go` line 128 at r20](https://reviewable.io/reviews/cockroachdb/cockroach/86565#-NAEvc0X5zN2Jp8X_ruM:-NAFnWbkEft0ZtSZkJU5:br6bymy) ([raw file](https://github.com/cockroachdb/cockroach/blob/2ee1f7caa4f2f189daabfb3318704c2ec4dbcb22/pkg/kv/kvserver/batcheval/cmd_delete_range.go#L128)):*
<details><summary><i>Previously, erikgrinaker (Erik Grinaker) wrote…</i></summary><blockquote>

Range tombstones are non-transactional, so this can only be `h.Timestamp`.
</blockquote></details>

It is also consistent with what we pass to actual delete operation this way.


<!-- Sent from Reviewable.io -->

erikgrinaker commented on pkg/kv/kvserver/client_split_test.go:
```diff
@@ -1951,6 +1951,68 @@ func TestStoreSplitGCThreshold(t *testing.T) {
 	repl.AssertState(ctx, store.Engine())
 }
 
+func TestStoreSplitGCHint(t *testing.T) {
```
Let's do a merge test too -- both when the LHS and RHS match, and when they don't (e.g. different value or missing value).

erikgrinaker reviewed (approved):
Seems good overall. I'll let @tbg have a closer look at the migration concerns, since I don't have a full picture of the pitfalls here.

It might be worth spinning up a mixed-version cluster, run a workload that does a bunch of splits and merges, and then to a full consistency check of all ranges, just to make sure we're not getting any divergence.

Oh, and add on a commit to set this in `gcjob.deleteAllSpanData()`.

tbg reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

instead of manual testing, could you add a step around here

https://github.com/cockroachdb/cockroach/blob/3525705b50b74c208df5d99228d9975cfac5e87a/pkg/cmd/roachtest/tests/versionupgrade.go#L146

that performs a few (random) splits and immediately merges some of them back (Like do three splits, merge the last two)? Can probably just add something into `versionUpgradeTestFeatures` https://github.com/cockroachdb/cockroach/blob/3525705b50b74c208df5d99228d9975cfac5e87a/pkg/cmd/roachtest/tests/versionupgrade.go#L49 to that effect:

```
	stmtFeatureTest("Split-then-merge", v201, `
CREATE DATABASE IF NOT EXISTS splitmerge;
CREATE TABLE splitmerge.t (id INT PRIMARY KEY);
ALTER TABLE splitmerge.t SPLIT AT VALUES(1,2,3); -- probably doing this somewhat wrong
ALTER TABLE splitmerge.t UNSPLIT AT VALUES(2,3); -- probably doing this somewhat wrong
`),
```

That will give us some confidence that things don't go sideways. We may also want this diff:

```go
diff --git a/pkg/cmd/roachtest/tests/versionupgrade.go b/pkg/cmd/roachtest/tests/versionupgrade.go
index ccae65f35f..8c414fba43 100644
--- a/pkg/cmd/roachtest/tests/versionupgrade.go
+++ b/pkg/cmd/roachtest/tests/versionupgrade.go
@@ -138,7 +138,9 @@ func runVersionUpgrade(ctx context.Context, t test.Test, c cluster.Cluster) {
 		// and finalizing on the auto-upgrade path.
 		preventAutoUpgradeStep(1),
 		// Roll nodes forward.
-		binaryUpgradeStep(c.All(), ""),
+		binaryUpgradeStep(c.Node(1), ""),
+		testFeaturesStep,
+		binaryUpgradeStep(c.Range(2, c.Spec().NodeCount), ""),
 		testFeaturesStep,
 		// Run a quick schemachange workload in between each upgrade.
 		// The maxOps is 10 to keep the test runtime under 1-2 minutes.
```

so that we have a chance to truly exercise this in the case where some nodes are new and some are old, but I think if this causes unrelated trouble we can kick it down the road.

> Reviewed 17 of 17 files at r25, 8 of 8 files at r26, 9 of 9 files at r27, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/86565#-:-NAJW9T13w0ljDfebxnz:b-p72u8z)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @aliher1911)

___
*[`pkg/kv/kvserver/stateloader/stateloader.go` line 308 at r26](https://reviewable.io/reviews/cockroachdb/cockroach/86565#-NAJVeVM9YNiJ2_KRujE:-NAJVeVM9YNiJ2_KRujF:bz6m4mm) ([raw file](https://github.com/cockroachdb/cockroach/blob/95ef66025dfd316ac9def8ad879e846b86eb1e3d/pkg/kv/kvserver/stateloader/stateloader.go#L308)):*
> ```Go
> 	hint *roachpb.GCHint,
> 	gcHintEnabled bool,
> ) (bool, error) {
> ```

`(updated bool, _ error)`


<!-- Sent from Reviewable.io -->

tbg reviewed (approved):
NB: not at all against manual testing. Just if we can spend the same time on augmenting the automatic test, it will likely give us more confidence. But if this is too expensive, definitely a manual test + then merge.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/86565#-:-NAJY-U_1UNSQF-Bq8--:bsonk88)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @aliher1911)



<!-- Sent from Reviewable.io -->

aliher1911 committed:
roachtest: versionupgrade to check range splits and merges

This commit adds a test that peforms range split and merge during
cluster upgrade. This is covering test cases where raft state gets
additional information in new versions and we must ensure that
replicas don't diverge in the process of upgrade.

Release justification: Commit adds extra tests only.
Release note: None

tbg reviewed (approved):
:passenger_ship: :it: 

> Reviewed 19 of 19 files at r28, 8 of 8 files at r29, 10 of 10 files at r30, 2 of 2 files at r31, all commit messages.
> Dismissed @erikgrinaker from a discussion.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/86565#-:-NAO5sDp2xHq8TutQuNK:b-p6wpwx)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @aliher1911)



<!-- Sent from Reviewable.io -->

aliher1911 commented:
For posterity, I tried testing manually with splits and merges initiated by "new" and "old" versions during upgrade in a mixed version cluster waited for it to sit after that for consistency checks to run and it didn't cause any trouble. So assuming all good on this front.

aliher1911 commented:
bors r=tbg,erikgrinaker

aliher1911 commented:
Touches https://github.com/cockroachdb/cockroach/issues/86550
